pkg/registry/generic/etcd/etcd_test.go

```diff
@@ -343,254 +265,106 @@ func TestEtcdUpdate(t *testing.T) {
-		R: &etcd.Response{},
-		E: tools.EtcdErrorNotFound,
+	// Test2 createIfNotFound and verify
+	registry.UpdateStrategy.(*testRESTStrategy).allowCreateOnUpdate = true
+	if !updateAndVerify(t, testContext, registry, podA) {
+		t.Errorf("Unexpected error updating podA")
 	}
+	registry.UpdateStrategy.(*testRESTStrategy).allowCreateOnUpdate = false
 
-	testContext := api.WithNamespace(api.NewContext(), "test")
+	// Test3 outofDate
+	_, _, err = registry.Update(testContext, podAWithResourceVersion)
+	if !errors.IsConflict(err) {
+		t.Errorf("Unexpected error: %v", err)
+	}
 
-	table := map[string]struct {
-		existing                 tools.EtcdResponseWithError
-		expect                   tools.EtcdResponseWithError
-		toUpdate                 runtime.Object
-		allowCreate              bool
-		allowUnconditionalUpdate bool
-		objOK                    func(obj runtime.Object) bool
-		errOK                    func(error) bool
-	}{
-		"normal": {
-			existing: nodeWithPodA,
-			expect:   nodeWithPodB,
-			toUpdate: podCopy(podB),
-			errOK:    func(err error) bool { return err == nil },
-		},
-		"notExisting": {
-			existing: emptyNode,
-			expect:   emptyNode,
-			toUpdate: podCopy(podA),
-			errOK:    func(err error) bool { return errors.IsNotFound(err) },
-		},
-		"createIfNotFound": {
-			existing:    emptyNode,
-			toUpdate:    podCopy(podA),
-			allowCreate: true,
-			objOK:       hasCreated(t, podA),
-			errOK:       func(err error) bool { return err == nil },
-		},
-		"outOfDate": {
-			existing: newerNodeWithPodA,
-			expect:   newerNodeWithPodA,
-			toUpdate: podCopy(podB),
-			errOK:    func(err error) bool { return errors.IsConflict(err) },
-		},
-		"unconditionalUpdate": {
-			existing:                 nodeWithPodAWithResourceVersion,
-			allowUnconditionalUpdate: true,
-			toUpdate:                 podCopy(podA),
-			objOK:                    func(obj runtime.Object) bool { return true },
-			errOK:                    func(err error) bool { return err == nil },
-		},
+	// Test4 normal update and verify
+	if !updateAndVerify(t, testContext, registry, podB) {
+		t.Errorf("Unexpected error updating podB")
 	}
 
-	for name, item := range table {
-		fakeClient, registry := NewTestGenericEtcdRegistry(t)
-		registry.UpdateStrategy.(*testRESTStrategy).allowCreateOnUpdate = item.allowCreate
-		registry.UpdateStrategy.(*testRESTStrategy).allowUnconditionalUpdate = item.allowUnconditionalUpdate
-		path := etcdtest.AddPrefix("pods/foo")
-		fakeClient.Data[path] = item.existing
-		obj, _, err := registry.Update(testContext, item.toUpdate)
-		if !item.errOK(err) {
-			t.Errorf("%v: unexpected error: %v", name, err)
-		}
+	// Test5 unconditional update
+	// NOTE: The logic for unconditional updates doesn't make sense to me, and imho should be removed.
+	// doUnconditionalUpdate := resourceVersion == 0 && e.UpdateStrategy.AllowUnconditionalUpdate()
+	// ^^ That condition can *never be true due to the creation of root objects.
```

Comment:
default creation of /registry/pods means that RV=2.. So the unconditional update logic will never be met.  (resourceVersion == 0) < never happens.